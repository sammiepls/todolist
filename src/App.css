.App {
  --space: 1.25rem;
  --half-space: calc(var(--space) / 2);
  --black: hsl(0, 0%, 3%);
  --white: hsl(0, 0%, 96%);
  --gray: hsl(0, 0%, 63%);
  --light-gray: hsl(0, 0%, 83%);
  --red: hsl(0, 72%, 49%);
  --primary: hsl(208, 76%, 61%);
  --border: 3px;
  --border-thin: 1px;
  --size-1: 1rem;
  --size-2: 2.42rem;

  color: var(--black);
  background: var(--white);
  display: flex;
  flex-flow: column;
}

.App-header {
  padding: var(--space);
  border-bottom: var(--border-thin) solid var(--light-gray);
}

.App-header > * {
  margin: 0;
}

.App-form {
  padding: var(--space);
  display: flex;
}

.App-form label span {
  display: none;
}


.App-form label {
  flex: 1;
  display: flex;
}

.App-form input {
  margin: 0;
  padding: 0 var(--half-space);
  flex: 1;
  border: none;
  border-bottom: var(--border) solid var(--primary);
}

.App-form button,
.Todo button {
  border: none;
  background: var(--primary);
  color: var(--white);
  padding: var(--half-space);
}

.App-form button {
  margin: 0;
}

.Todo-list {
  list-style: none;
  padding: 0;
  margin: 0;
}

.Todo {
  display: flex;
  align-items: center;
  margin: var(--space);
  padding: 0 0 0 var(--half-space);
}

.Todo:hover {
  background: var(--light-gray);
}

.Todo[data-done="true"] .Todo-title {
  text-decoration: line-through;
  color: var(--gray);
}

.Todo button {
  margin-left: auto;
}

.Todo-title {
  flex: 1;
  cursor: pointer;
}

.Todo-placeholder {
  padding: var(--space);
  font-size: var(--size-2);
  font-weight: bold;
  margin: 0;
  color: var(--gray);
}

.Error {
  margin: var(--space);
  color: var(--red);
}
