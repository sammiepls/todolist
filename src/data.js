const todos = [
  {
    title: "Do something",
    done: false,
    id: 'todo_1'
  },
  {
    title: "Do nothing",
    done: false,
    id: 'todo_2'
  },
  {
    title: "Do everything",
    done: false,
    id: 'todo_3'
  }
];

const notification = {
  msg: "Test",
  class: "test"
}

export { todos, notification };
