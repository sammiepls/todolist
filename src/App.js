import React, { Component } from 'react';
import TodoList from './components/TodoList'
import './App.css';

class App extends Component {
  constructor(props) {
    super(props)
    this.state = {
      todos: [],
      notification: {},
      todo: ""
    }
  }

  toggleDone = (id) => {
    const prevTodos = this.state.todos
    this.setState({
      todos: prevTodos.map(todo => {
        if (todo.id === id) {
          todo.done = !todo.done
        }

        return todo 
      })
    })
  }

  deleteTodo = (id) => {
    const prevTodos = [...this.state.todos]
    this.setState({
      todos: prevTodos.filter((todo) => todo.id !== id)
    })
    this.setNotification({ msg: "Deleted successfully", class: "success" })
  }

  handleSubmit = (e) => {
    e.preventDefault()
    // Validate and sanitize
    let title = this.state.todo.trim()
    if (title.length <= 0) {
      this.setNotification({ msg: "Todo can not be blank", class: "error" })
      return false
    }
    
    // Add to state
    this.setState((state) => {
      return {
        todos: [...state.todos, {
          title,
          done: false,
          id: `todo_${new Date().getTime()}`
        }]
      }
    })
    
    // Clear input
    this.clearNotification()
  }
  
  handleInputChange = (e) => this.setState({ todo: e.target.value })

  get incompleteTodos() {
    return this.state.todos.filter(todo => todo.done === false)
  }

  get completedTodos() {
    return this.state.todos.filter(todo => todo.done)
  }

  setNotification = (notification) => {
    this.setState({ 
      notification
    })
  }

  clearNotification = (e) => this.setState({ notification: {} })

  render() {
    return (
      <React.Fragment>
        { Object.keys(this.state.notification).length > 0 && (<p className={ "App-notification " + (this.state.notification.class ? this.state.notification.class : '') } onClick={ this.clearNotification }>{ this.state.notification.msg }</p>) }
        <header className="App-header">
          <h1 className="App-heading">To do</h1>
        </header>
        <main>
          <h2 className="App-heading">Incomplete</h2>
          <TodoList todos={ this.incompleteTodos } 
            onDeleteTodo={ this.deleteTodo }
            onToggleDone={ this.toggleDone }></TodoList>

          { this.completedTodos.length > 0 && (
            <React.Fragment>
            <h2 className="App-heading">Completed</h2>
            <TodoList todos={ this.completedTodos } 
              onDeleteTodo={ this.deleteTodo }
              onToggleDone={ this.toggleDone }></TodoList>
            </React.Fragment>
          )}
        </main>
        <footer>
          <form className="App-form" onSubmit={ this.handleSubmit }>
            <label>
              <span>title</span>
              <input type="text" name="title" onChange={ this.handleInputChange } />
            </label>
            <button>Add</button>
          </form>
        </footer>
      </React.Fragment>
    );
  }
}

export default App;
