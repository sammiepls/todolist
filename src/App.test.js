import React from 'react'
import App from './App'
import { shallow } from 'enzyme'
import renderer from 'react-test-renderer'
import { todos, notification } from './data'


const defaultState = { todos, notification: {}, todo: "" }

describe('App.js', () => {
  const wrapper = shallow(<App/>)

  it('renders without crashing', () => {
    const wrapper = renderer.create(<App />)
    expect(wrapper.toJSON()).toMatchSnapshot()
  })

  it('should have correct state', () => {
    expect(wrapper.state()).toEqual({ todos: [], notification: {}, todo: "" })
  })

  describe('Notification', () => {
    const className = ".App-notification"

    it('should be able to display notification', () => {
      const Notification = wrapper.find(className)

      // Check notification is clear
      expect(wrapper.state().notification).toEqual({})
      // Set test notification
      wrapper.instance().setNotification(notification)
      // Check for notification
      expect(wrapper.find(className)).toHaveLength(1)
    })

    it('should be able to clear notification', () => {
      wrapper.setState({ notification })
      // Check if notification is visible
      expect(wrapper.find(className)).toHaveLength(1)
      // Trigger close
      wrapper.find(className).simulate('click')
      expect(wrapper.find(className)).toHaveLength(0)
    })
  })

  describe('Todo', () => {
    beforeEach(() => {
      wrapper.setState(defaultState)
    })

    it('should be able to delete todo', () => {
      const todoId = todos[0].id
      expect(wrapper.state().todos).toHaveLength(todos.length)

      wrapper.instance().deleteTodo(todoId)
      expect(wrapper.state().todos).toHaveLength(todos.length - 1)
    })

    it('should be able to toggle todo done status', () => {
      const todoId = todos[0].id
      wrapper.instance().toggleDone(todoId)
      expect(wrapper.state().todos.filter(t => t.id === todoId)[0].done).toBe(true)
      wrapper.instance().toggleDone(todoId)
      expect(wrapper.state().todos.filter(t => t.id === todoId)[0].done).toBe(false)
    })
  })

  describe('Form', () => {
    const fakeEvent = { preventDefault: () => 'I is fake event' }
    const form = wrapper.find('.App-form')

    it('should not be able to add empty todo', () => {
      form.simulate('submit', fakeEvent)
      expect(wrapper.find('.App-notification').text()).toBe("Todo can not be blank")
    })

    it('should be able to add todo', () => {
      const testTask = "Test task"
      form.find('input').simulate('change', { target: { value: testTask } })
      form.simulate('submit', fakeEvent)

      expect(wrapper.state().todos).toHaveLength(todos.length + 1)
    })
  })
})
