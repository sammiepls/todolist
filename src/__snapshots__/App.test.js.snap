// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App.js renders without crashing 1`] = `
Array [
  <header
    className="App-header"
  >
    <h1
      className="App-heading"
    >
      To do
    </h1>
  </header>,
  <main>
    <h2
      className="App-heading"
    >
      Incomplete
    </h2>
    <p
      className="Todo-placeholder"
    >
      No items found, add an item.
    </p>
  </main>,
  <footer>
    <form
      className="App-form"
      onSubmit={[Function]}
    >
      <label>
        <span>
          title
        </span>
        <input
          name="title"
          onChange={[Function]}
          type="text"
        />
      </label>
      <button>
        Add
      </button>
    </form>
  </footer>,
]
`;
